Userland/Libraries/LibWeb/HTML/CloseWatcher.cpp

```diff
@@ -0,0 +1,174 @@
+/*
+ * Copyright (c) 2024, the Ladybird developers.
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include <AK/TypeCasts.h>
+#include <LibWeb/Bindings/CloseWatcherPrototype.h>
+#include <LibWeb/Bindings/Intrinsics.h>
+#include <LibWeb/DOM/Document.h>
+#include <LibWeb/DOM/EventDispatcher.h>
+#include <LibWeb/DOM/IDLEventListener.h>
+#include <LibWeb/HTML/CloseWatcher.h>
+#include <LibWeb/HTML/CloseWatcherManager.h>
+#include <LibWeb/HTML/EventHandler.h>
+#include <LibWeb/HTML/Window.h>
+
+namespace Web::HTML {
+
+JS_DEFINE_ALLOCATOR(CloseWatcher);
+
+// https://html.spec.whatwg.org/multipage/interaction.html#establish-a-close-watcher
+JS::NonnullGCPtr<CloseWatcher> CloseWatcher::establish(HTML::Window& window)
+{
+    // 1. Assert: window's associated Document is fully active.
+    VERIFY(window.associated_document().is_fully_active());
+
+    // 2. Let closeWatcher be a new close watcher
+    auto close_watcher = window.heap().allocate<CloseWatcher>(window.realm(), window.realm());
+
+    // 3. Let manager be window's associated close watcher manager
+    auto manager = window.close_watcher_manager();
+
+    // 4 - 6. Moved to CloseWatcherManager::add
+    manager->add(close_watcher);
+
+    // 7. Return close_watcher.
+    return close_watcher;
+}
+
+// https://html.spec.whatwg.org/multipage/interaction.html#dom-closewatcher
+WebIDL::ExceptionOr<JS::NonnullGCPtr<CloseWatcher>> CloseWatcher::construct_impl(JS::Realm& realm, CloseWatcherOptions const& options)
+{
+    // 1. If this's relevant global object's associated Document is not fully active, then return an "InvalidStateError" DOMException.
+    auto& window = verify_cast<HTML::Window>(realm.global_object());
+    if (!window.associated_document().is_fully_active())
+        return WebIDL::InvalidStateError::create(realm, "The document is not fully active."_fly_string);
+
+    // 2. Let close_watcher be the result of establishing a close watcher
+    auto close_watcher = establish(window);
+
+    // 3. If options["signal"] exists, then:
+    if (options.signal) {
+        // 3.1 If options["signal"]'s aborted, then destroy close_watcher.
+        if (options.signal->aborted())
+            close_watcher->m_is_active = false;
+        // TODO: Do I need to add the signal to the close watcher?
+        // 3.2 Add the following steps to options["signal"]:
+        options.signal->add_abort_algorithm([&close_watcher] {
+            // 3.2.1 Destroy closeWatcher.
+            close_watcher->m_is_active = false;
```

Comment:
Remember to swap to calling destroy once the issue has been worked out. Removing the VERIFY too